frontend/packages/dev-console/src/components/import/git/SourceSecretSelector.tsx

```diff
@@ -1,46 +1,84 @@
+import { FormGroup, ControlLabel } from 'patternfly-react';
+import { useFormikContext, FormikValues, useField } from 'formik';
+import { createModalLauncher } from '@console/internal/components/factory/modal';
+import {
+  withSecretForm,
+  SourceSecretForm,
+  SecretTypeAbstraction,
+} from '@console/internal/components/secrets/create-secret';
 import SourceSecretDropdown from '../../dropdown/SourceSecretDropdown';
-import CreateSourceSecret, { SecretType } from './CreateSourceSecret';
 
 const CREATE_SOURCE_SECRET = 'create-source-secret';
 
 interface SourceSecretSelectorProps {
-  namespace?: string;
-  sourceSecret: string;
-  selectedKey: string;
-  secretCredentials: { [key: string]: string };
-  onChange: (
-    name: string,
-    key: string,
-    authType?: string,
-    credentials?: { [key: string]: string },
-  ) => void;
+  namespace: string;
 }
 
-const SourceSecretSelector: React.FC<SourceSecretSelectorProps> = ({
-  sourceSecret,
+interface CreateSourceSecretModalProps {
+  cancel: (e: MouseEvent) => void;
+  close: () => void;
+  onSave?: (name: string) => void;
+  namespace: string;
+}
+
+const CreateSourceSecretModal: React.FC<CreateSourceSecretModalProps> = ({
+  close,
   namespace,
-  secretCredentials,
-  selectedKey,
-  onChange,
+  onSave,
 }) => {
-  const onDropdownChange = (secretName: string, key: string) => {
-    if (key === CREATE_SOURCE_SECRET) {
-      onChange('', key, SecretType.basicAuth, {});
-    } else {
-      onChange(secretName, key);
-    }
+  const CreateSourceSecretForm = withSecretForm(SourceSecretForm, true);
+  const onClose = () => {
+    close();
+    return null;
   };
 
-  const onInputChange: React.ReactEventHandler<HTMLInputElement> = (event) => {
-    onChange(event.currentTarget.value, selectedKey, SecretType.basicAuth, secretCredentials);
+  const save = (name: string) => {
+    close();
+    onSave(name);
   };
+  return (
+    <CreateSourceSecretForm
+      onCancel={onClose}
+      onSave={save}
+      fixed={{ metadata: { namespace } }}
+      secretTypeAbstraction={SecretTypeAbstraction.source}
+      explanation="Source secrets let you authenticate against a Git server."
+      titleVerb="Create"
+      isCreate
+    />
+  );
+};
+
+const sourceSecretModalLauncher = createModalLauncher<CreateSourceSecretModalProps>(
+  CreateSourceSecretModal,
+);
 
+const SourceSecretSelector: React.FC<SourceSecretSelectorProps> = ({ namespace }) => {
+  const [selectedKey] = useField('git.secret.selectedKey');
+  const { setFieldValue } = useFormikContext<FormikValues>();
+
+  let secretName: string;
+
+  const onSave = (name: string) => {
+    secretName = name;
+  };
+
+  const onDropdownChange = async (key: string) => {
+    if (key === CREATE_SOURCE_SECRET) {
+      secretName='';
+      setFieldValue('git.secret.selectedKey', key);
+      await sourceSecretModalLauncher({ namespace, onSave }).result.then(() => {
+        setFieldValue('git.secret.selectedKey', secretName);
```

Comment:
Move this to `onSave` so then you can remove your local variable `secretName`

Reply:
@christianvogt  When user clicks on `Create new secret` and then clicks `cancel` or clicks `outside the modal` to close it, the dropdown title should show `Select New Secret`. So the `setFieldValue` inside the promise also  supports these two scenarios apart from saving the new secret. In occurrence of any of the above two scenarios the secretName is empty and that is being set as the value of the selectedkey

Reply:
If the user has a secret selected, then clicks the `create new secret` action, followed by cancelling the modal, I would expect the previously selected secret to remain selected.
Similarly I also expect if there is no selection and the placeholder text is shown to still be shown if they cancel the modal.

Reply:
Done